File: modules/sql/app/model/querycorrectors/change/ChangeCorrector.java
package model.querycorrectors.change;

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Collections;
import java.util.List;

import model.CorrectionException;
import model.StringConsts;
import model.exercise.SqlExercise;
import model.matching.Match;
import model.matching.MatchingResult;
import model.querycorrectors.QueryCorrector;
import model.querycorrectors.SqlExecutionResult;
import model.querycorrectors.SqlResult;
import model.sql.SqlQueryResult;
import net.sf.jsqlparser.schema.Column;
import play.Logger;
import play.db.Database;

public abstract class ChangeCorrector<Q extends net.sf.jsqlparser.statement.Statement>
    extends QueryCorrector<Q, Column> {
  
  public ChangeCorrector(String theQueryType) {
    super(theQueryType);
  }
  
  private static void runUpdate(Connection connection, String query) {
    try(Statement statement = connection.createStatement()) {
      statement.executeUpdate(query);
    } catch (SQLException e) {
      Logger.error("There has been an error running an sql query: \"" + query + "\"", e);
    }
  }
  
  private static SqlQueryResult runValidationQuery(Connection connection, String query) throws CorrectionException {
    try(Statement statement = connection.createStatement()) {
      return new SqlQueryResult(statement.executeQuery(query));
    } catch (SQLException e) {
      throw new CorrectionException(query, "There has been an error running an sql query", e);
    }
  }
  
  private SqlQueryResult getResultSet(Q statement, Connection connection, String validation)
      throws CorrectionException, SQLException {
    
    runUpdate(connection, statement.toString());
    
    SqlQueryResult result = runValidationQuery(connection, validation);
    
    connection.rollback();
    
    return result;
  }
  
  @Override
  protected SqlExecutionResult executeQuery(Database database, Q userStatement, Q sampleStatement, SqlExercise exercise)
      throws CorrectionException {
    try(final Connection connection = database.getConnection()) {
      
      connection.setCatalog(exercise.scenario.getShortName());
      connection.setAutoCommit(false);
      
      final String validation = StringConsts.SELECT_ALL_DUMMY + getTableNames(sampleStatement).get(0);
      
      final SqlQueryResult userResult = getResultSet(userStatement, connection, validation);
      
      final SqlQueryResult sampleResult = getResultSet(sampleStatement, connection, validation);
      
      return new SqlExecutionResult(userResult, sampleResult);
    } catch (SQLException e) {
      throw new CorrectionException(userStatement.toString(), "There has been an error running a statement", e);
    }
  }
  
  @Override
  protected SqlResult<Column> instantiateResult(String learnerSolution) {
    return new SqlResult<>(learnerSolution);
  }
  
  @Override
  protected List<MatchingResult<? extends Object, ? extends Match<? extends Object>>> makeOtherComparisons(Q userQ,
      Q sampleQ) {
    return Collections.emptyList();
  }
  
}

File: modules/sql/app/model/querycorrectors/ChangeCorrector.scala
package model.querycorrectors

import java.sql.Connection

import model.CommonUtils.cleanly
import model.exercise.SqlExercise
import model.matching.MatchingResult
import model.sql.SqlQueryResult
import model.{CorrectionException, StringConsts}
import net.sf.jsqlparser.JSQLParserException
import net.sf.jsqlparser.expression.Expression
import net.sf.jsqlparser.parser.CCJSqlParserUtil
import net.sf.jsqlparser.schema.Table
import net.sf.jsqlparser.statement.select.OrderByElement
import play.db.Database

import scala.collection.JavaConverters._
import scala.util.Try

abstract class ChangeCorrector(queryType: String) extends QueryCorrector(queryType) {

  def runUpdate(conn: Connection, query: String): Try[Int] = cleanly(conn.createStatement)(_.close)(_.executeUpdate(query))

  def runValidationQuery(conn: Connection, query: String): Try[SqlQueryResult] =
    cleanly(conn.createStatement)(_.close)(s => new SqlQueryResult(s.executeQuery(query)))

  def getResultSet(statement: Q, connection: Connection, validation: String): Try[SqlQueryResult] = {
    runUpdate(connection, statement.toString)
    val result = runValidationQuery(connection, validation)
    connection.rollback()
    result
  }

  override def executeQuery(db: Database, userQ: Q, sampleQ: Q, exercise: SqlExercise): Try[SqlExecutionResult] =
    cleanly(db.getConnection)(_.close)(connection => {
      connection.setCatalog(exercise.scenario.shortName)
      connection.setAutoCommit(false)

      val validation = StringConsts.SELECT_ALL_DUMMY + getTableNames(sampleQ).head

      new SqlExecutionResult(getResultSet(userQ, connection, validation).get, getResultSet(sampleQ, connection, validation).get)
    })

  override def compareGroupByElements(plainUserQuery: Q, plainSampleQuery: Q): Option[MatchingResult[Expression, GroupByMatch]] = None

  override def compareOrderByElements(plainUserQuery: Q, plainSampleQuery: Q): Option[MatchingResult[OrderByElement, OrderByMatch]] = None
}

object InsertCorrector extends ChangeCorrector("INSERT") {

  import net.sf.jsqlparser.statement.insert.Insert

  type Q = Insert

  override def getColumnWrappers(query: Q): List[ColumnWrapper] = List.empty

  override def getTableNames(query: Q) = List(query.getTable.getName)

  override def getTables(query: Q) = List(query.getTable)

  override def getWhere(query: Q): Expression = null

  def parseStatement(statement: String): Insert = try {
    val parsed = CCJSqlParserUtil.parse(statement)
    parsed match {
      case q: Q => q
      case _ => throw new CorrectionException(statement, "Das Statement war vom falschen Typ! Erwartet wurde " + queryType + "!")
    }
  } catch {
    case e: JSQLParserException => throw new CorrectionException(statement, "Es gab einen Fehler beim Parsen des Statements: " + statement, e)
  }

}

object DeleteCorrector extends ChangeCorrector("DELETE") {

  import net.sf.jsqlparser.statement.delete.Delete

  type Q = Delete

  override def getColumnWrappers(query: Q): List[ColumnWrapper] = List.empty

  override def getTableNames(query: Q) = List(query.getTable.getName)

  override def getTables(query: Q): List[Table] = query.getTables.asScala.toList

  override def getWhere(query: Q): Expression = query.getWhere

  def parseStatement(statement: String): Delete = try {
    val parsed = CCJSqlParserUtil.parse(statement)
    parsed match {
      case q: Q => q
      case _ => throw new CorrectionException(statement, "Das Statement war vom falschen Typ! Erwartet wurde " + queryType + "!")
    }
  } catch {
    case e: JSQLParserException => throw new CorrectionException(statement, "Es gab einen Fehler beim Parsen des Statements: " + statement, e)
  }

}

object UpdateCorrector extends ChangeCorrector("UPDATE") {

  import net.sf.jsqlparser.statement.update.Update

  type Q = Update

  override def getColumnWrappers(query: Q): List[ColumnWrapper] = query.getColumns.asScala.map(ColumnWrapper.wrap).toList

  override def getTableNames(query: Q): List[String] = query.getTables.asScala.map(_.getName).toList

  override def getTables(query: Q): List[Table] = query.getTables.asScala.toList

  override def getWhere(query: Q): Expression = query.getWhere

  def parseStatement(statement: String): Update = try {
    val parsed = CCJSqlParserUtil.parse(statement)
    parsed match {
      case q: Q => q
      case _ => throw new CorrectionException(statement, "Das Statement war vom falschen Typ! Erwartet wurde " + queryType + "!")
    }
  } catch {
    case e: JSQLParserException => throw new CorrectionException(statement, "Es gab einen Fehler beim Parsen des Statements: " + statement, e)
  }

}
